File: polocloud-common/src/main/java/dev/httpmarco/polocloud/common/OS.java
package dev.httpmarco.polocloud.common;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.experimental.Accessors;

@Getter
@Accessors(fluent = true)
@AllArgsConstructor
public enum OS {

    WINDOWS(";"),
    LINUX(":"),
    MACOS(":"),
    UNKNOWN(":");

    private final String processSeparator;

    public static OS detect() {
        String osName = System.getProperty("os.name").toLowerCase();

        if (osName.contains("win")) {
            return WINDOWS;
        } else if (osName.contains("nux") || osName.contains("nix")) {
            return LINUX;
        } else if (osName.contains("mac")) {
            return MACOS;
        } else {
            return UNKNOWN;
        }
    }
}

File: common/src/main/kotlin/dev/httpmarco/polocloud/common/os/OS.kt
package dev.httpmarco.polocloud.common.os

enum class OS(
    val nativeExecutableSuffix: String?,
    val shellPrefix: Array<String>,
    val currentDirectoryPrefix: String?
) {
    WIN("exe",  arrayOf("cmd", "/c"), null),
    LINUX(null, arrayOf("sh", "-c"), "./"),
    MACOS(null, arrayOf("sh", "-c"), "./"),
    UNKNOWN(null, emptyArray(), null);

    fun executableCurrentDirectoryCommand(filename: String): Array<String> {
        return arrayOf(*shellPrefix, "${currentDirectoryPrefix ?: ""}$filename")
    }
}